File: generators/dockerfiles/workers/worker.py
"""Worker for running Python scripts in client docker container."""

import logging
import os
import runpy
import sys
import time
import traceback

SCRIPT_PATH = "/tmp/invocation.py"
TRIGGER_PATH = "/tmp/run.trigger"
POLL_INTERVAL = 0.05  # seconds

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(levelname)s %(message)s",
    handlers=[logging.StreamHandler(sys.stdout)],
)

logger = logging.getLogger(__name__)


def run_script():
    logger.info(f"Executing {SCRIPT_PATH}")

    try:
        runpy.run_path(SCRIPT_PATH, run_name="__main__")
    except Exception:
        logger.error("Script failed:")
        traceback.print_exc()
    else:
        logger.info("Script completed successfully")


def main():
    logger.info("Worker started.")

    while True:
        if os.path.exists(TRIGGER_PATH):
            # Consume the trigger
            os.remove(TRIGGER_PATH)

            if os.path.exists(SCRIPT_PATH):
                run_script()
            else:
                logger.error(f"{SCRIPT_PATH} does not exist.")

        time.sleep(POLL_INTERVAL)


if __name__ == "__main__":
    main()

File: generators/dockerfiles/workers/csharp-worker/worker.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Scripting;
using Microsoft.CodeAnalysis.Scripting;
using System;
using System.IO;
using System.Threading;

const string AppRoot = "/app";
const string ScriptPath = $"{AppRoot}/invocation.csx";
const string TriggerPath = "/tmp/run.trigger";
const int PollIntervalMs = 50;

Console.WriteLine("Worker started.");
Console.WriteLine($"Working directory: {Environment.CurrentDirectory}");
Console.WriteLine($"Script path: {ScriptPath}");
Console.WriteLine($"Script exists: {File.Exists(ScriptPath)}");

while (true)
{
    if (File.Exists(TriggerPath))
    {
        try
        {
            File.Delete(TriggerPath);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Failed to remove trigger: {ex}");
            Thread.Sleep(PollIntervalMs);
            continue;
        }

        if (File.Exists(ScriptPath))
        {
            await RunScriptAsync();
        }
        else
        {
            Console.Error.WriteLine($"{ScriptPath} does not exist.");
        }
    }

    Thread.Sleep(PollIntervalMs);
}

static async Task RunScriptAsync()
{
    Console.WriteLine($"Executing {ScriptPath}");

    try
    {
        if (!File.Exists(ScriptPath))
        {
            throw new FileNotFoundException(
                "Script file was not found.",
                ScriptPath);
        }

        var code = await File.ReadAllTextAsync(ScriptPath);

        var options = ScriptOptions.Default
            .WithLanguageVersion(LanguageVersion.Latest)
            .WithImports(
                "System",
                "System.IO",
                "System.Linq",
                "System.Collections.Generic",
                "System.Threading",
                "System.Threading.Tasks"
            )
            .AddReferences(
                MetadataReference.CreateFromFile(
                    "/usr/share/dotnet/shared/Microsoft.AspNetCore.App/10.0.8/Microsoft.Extensions.Logging.Abstractions.dll"
                )
            )
            .WithFilePath(ScriptPath);

        await CSharpScript.RunAsync(code, options);

        Console.WriteLine("Script completed successfully");
    }
    catch (Exception ex)
    {
        Console.Error.WriteLine("Script failed:");
        Console.Error.WriteLine(ex);
    }
}
